dtm0/dtx.c

```diff
@@ -0,0 +1,540 @@
+		    uint32_t             nr_pa)
+{
+	M0_PRE(dtx != NULL);
+	M0_ENTRY("dtx=%p", dtx);
+	return M0_RC(m0_dtm0_tx_desc_init(&dtx->dd_txd, nr_pa));
+}
+
+static void dtx_fop_assign(struct m0_dtm0_dtx  *dtx,
+			   uint32_t             pa_idx,
+			   const struct m0_fop *pa_fop)
+{
+	M0_PRE(dtx != NULL);
+	M0_PRE(pa_idx < dtx->dd_txd.dtd_ps.dtp_nr);
+
+	M0_ENTRY("dtx=%p, pa=%" PRIu32 ", fop=%p", dtx, pa_idx, pa_fop);
+
+	(void) pa_idx;
+
+	/*
+	 * TODO:REDO: On the DTM side we should enforce the requirement
+	 * described at m0_dtm0_dtx::dd_op.
+	 * At this moment we silently ignore this as well as anything
+	 * related directly to REDO use-cases.
+	 */
+	if (dtx->dd_fop == NULL) {
+		dtx->dd_fop = pa_fop;
+	}
+
+	M0_LEAVE();
+}
+
+
+static int dtx_fid_assign(struct m0_dtm0_dtx  *dtx,
+			  uint32_t             pa_idx,
+			  const struct m0_fid *pa_sfid)
+{
+	struct m0_dtm0_tx_pa   *pa;
+	struct m0_reqh         *reqh;
+	struct m0_conf_cache   *cache;
+	struct m0_conf_obj     *obj;
+	struct m0_conf_process *proc;
+	struct m0_fid           rdtms_fid;
+	int                     rc;
+
+	M0_ENTRY("dtx=%p", dtx);
+
+	M0_PRE(dtx != NULL);
+	M0_PRE(pa_idx < dtx->dd_txd.dtd_ps.dtp_nr);
+	M0_PRE(m0_fid_is_valid(pa_sfid));
+
+	/* XXX: Should we lock or release any conf objects or the cache? */
+
+	reqh = dtx->dd_dtms->dos_generic.rs_reqh;
+	cache = &m0_reqh2confc(reqh)->cc_cache;
+
+	obj = m0_conf_cache_lookup(cache, pa_sfid);
+	if (obj == NULL)
+		return M0_ERR_INFO(-ENOENT,
+				   "User service is not in the conf cache.");
+
+	obj = m0_conf_obj_grandparent(obj);
+	if (obj == NULL)
+		return M0_ERR_INFO(-ENOENT, "Process the service belongs to "
+		       "is not a part of the conf cache.");
+
+	proc = M0_CONF_CAST(obj, m0_conf_process);
+	M0_ASSERT_INFO(proc != NULL, "The grandparent is not a process?");
+
+	rc = m0_conf_process2service_get(&reqh->rh_rconfc.rc_confc,
+					 &proc->pc_obj.co_id, M0_CST_DTM0,
+					 &rdtms_fid);
+	if (rc != 0)
+		return M0_ERR_INFO(rc, "Cannot find remote DTM service on"
+				   " the remote process that runs"
+				   " this user service.");
+
+	pa = &dtx->dd_txd.dtd_ps.dtp_pa[pa_idx];
+	M0_ASSERT_INFO(M0_IS0(pa), "FID cannot be re-assigned.");
+	M0_ASSERT(m0_fid_is_valid(&rdtms_fid));
+
+	pa->p_fid = rdtms_fid;
+	M0_ASSERT(pa->p_state == M0_DTPS_INIT);
+	pa->p_state = M0_DTPS_INPROGRESS;
+
+	return M0_RC_INFO(0, "pa %" PRIu32 ": "
+			  FID_F " (User svc) => " FID_F " (DTM svc) ",
+			  pa_idx, FID_P(pa_sfid), FID_P(&rdtms_fid));
+}
+
+static int dtx_close(struct m0_dtm0_dtx *dtx)
+{
+	int rc;
+
+	M0_ENTRY("dtx=%p", dtx);
+
+	M0_PRE(dtx != NULL);
+	M0_PRE(m0_sm_group_is_locked(dtx->dd_sm.sm_grp));
+
+	/*
+	 * TODO:REDO: We may want to capture the fop contents here.
+	 * See ::fol_record_pack and ::m0_fop_encdec for the details.
+	 * At this moment we do not do REDO, so that it is safe to
+	 * avoid any actions on the fop here.
+	 */
+
+	rc = dtx_log_insert(dtx);
+	M0_ASSERT(rc == 0);
```

Comment:
error handling

Reply:
Will be handled during working on failure path.